### PR TITLE
External volume integration

### PR DESCRIPTION
**Pull request by philwinder**
Adds external volumes. Testing with rexray.

### CODE REVIEW

**Comment by philwinder** [comment]
Waiting to test on cluster, with rexray rc3. Previous versions don't work with Docker 1.10.

**Comment by philwinder** [comment]
Manually tested to be working, both docker mode and jar mode, using rexray version:
```
$ rexray version
Binary: /usr/bin/rexray
SemVer: 0.3.2-rc5
OsArch: Linux-x86_64
Branch: v0.3.2-rc5
Commit: 2533d497729d28c4982fb0112ea400b19a68532e
Formed: Thu, 25 Feb 2016 18:11:10 UTC
```
Note that the first time volumes were created, rexray timed out and returned a 503. This is a rexray bug. After a few retries, the executors were able to connect to the volumes.

**Comment by philwinder** [comment]
Jenkins pass
https://ci.infra.container-solutions.com/view/All/job/ElasticsearchPullRequests/291/console

File: docs/index.md [hunk 1]
```diff
@@ -158,6 +160,14 @@ Usage: (Options preceded by an asterisk are required) [options]
     --executorName
        The name given to the executor task.
        Default: elasticsearch-executor
+    --externalVolumeDriver
+       This defines the use of an external storage drivers to be used. By
+       default, elastic serch nodes will not be created with external volumes but rather
+       direct attached storage.
```
Comment (frankscholten) [inline docs/index.md]: Typo elastic serch

File: docs/index.md [hunk 2]
```diff
@@ -252,6 +262,117 @@ It is strongly recommended that you use the containerized version of Mesos Elast
...
+ssh -i $KEY ubuntu@$1 'echo { \"libraries\": [ { \"file\": \"/usr/lib/libmesos_dvdi_isolator-0.25.0.so\", \"modules\": [ { \"name\": \"com_emccode_mesos_DockerVolumeDriverIsolator\" } ] } ] } | sudo tee /usr/lib/dvdi-mod.json'
+ssh -i $KEY ubuntu@$1 'echo file:///usr/lib/dvdi-mod.json | sudo tee /etc/mesos-slave/modules'
+ssh -i $KEY ubuntu@$1 'echo com_emccode_mesos_DockerVolumeDriverIsolator | sudo tee /etc/mesos-slave/isolation'
+
+# Restart mesos slave to load new module
+ssh -i $KEY ubuntu@$1 'sudo service mesos-slave restart'
+```
+
 ### Data directory
 The ES node data can be written to a specific directory. If in docker mode, use the `--dataDir` option. If in jar mode, set the `path.data` option in your custom ES settings file.
```
Comment (frankscholten) [inline docs/index.md]: Perhaps we should move those scripts into separate files and reference them.

File: scheduler/src/main/java/org/apache/mesos/elasticsearch/scheduler/TaskInfoFactory.java [hunk 3]
```diff
@@ -190,14 +224,14 @@ private String resolveHostAddress(Protos.Offer offer, List<Integer> ports) {
...
         String address = configuration.getFrameworkFileServerAddress();
         if (address == null) {
             throw new NullPointerException("Webserver address is null");
         }
         String httpPath = address + "/get/" + Configuration.ES_TAR;
         String command = configuration.nativeCommand(args);
-        final Protos.Environment environment = Protos.Environment.newBuilder().addAllVariables(new ExecutorEnvironmentalVariables(configuration).getList()).build();
+        final Protos.Environment environment = Protos.Environment.newBuilder().addAllVariables(new ExecutorEnvironmentalVariables(configuration, lElasticSearchNodeId).getList()).build();
         final Protos.CommandInfo.Builder builder = Protos.CommandInfo.newBuilder()
                 .setShell(true)
```
Comment (frankscholten) [inline scheduler/src/main/java/org/apache/mesos/elasticsearch/scheduler/TaskInfoFactory.java]: Why is the variable called lElasticSearchNodeId

File: scheduler/src/main/java/org/apache/mesos/elasticsearch/scheduler/configuration/ExecutorEnvironmentalVariables.java [hunk 4]
```diff
@@ -35,6 +64,52 @@ public ExecutorEnvironmentalVariables(Configuration configuration) {
...
+        }
+        addToList(DVDI_VOLUME_CONTAINERPATH, CONTAINER_PATH_SETTINGS);
+
+        //sets the environment variables for to create and/or attach the data volume
+        //to the mesos containerizer
+        addToList(DVDI_VOLUME_DRIVER + "1", configuration.getExternalVolumeDriver());
+        addToList(DVDI_VOLUME_NAME + "1", sHostPathOrExternalVolumeForData);
+        if (configuration.getExternalVolumeOption() != null && configuration.getExternalVolumeOption().length() > 0) {
+            addToList(DVDI_VOLUME_OPTS + "1", configuration.getExternalVolumeOption());
+        }
```
Comment (frankscholten) [inline scheduler/src/main/java/org/apache/mesos/elasticsearch/scheduler/configuration/ExecutorEnvironmentalVariables.java]: Move these methods into Configuration. Makes them easier to test.

**Comment by philwinder** [comment]
Jenkins pass
https://ci.infra.container-solutions.com/view/Mesos%20Elasticsearch/job/ElasticsearchPullRequests/294/console

**Pull request merged**